enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,279 @@
+			return err
+		}
+		_, err = txn.Get([]byte(keyPrefix + sourceID))
+		return err
+	})
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Add adds the given suppressions to the repository
+func (b *Repository) Add(suppressions []model.Suppression, token []byte) error {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return badger.ErrDBClosed
+	}
+	wb := b.db.NewWriteBatch()
+	defer wb.Cancel()
+
+	for i := range suppressions {
+		suppression := suppressions[i]
+		keyPrefix := fmt.Sprintf("%s:%s:", suppression.WorkspaceID, suppression.UserID)
+		var keys []string
+		if len(suppression.SourceIDs) == 0 {
+			keys = []string{keyPrefix + model.Wildcard}
+		} else {
+			keys = make([]string, len(suppression.SourceIDs))
+			for i, sourceID := range suppression.SourceIDs {
+				keys[i] = keyPrefix + sourceID
+			}
+		}
+		for _, key := range keys {
+			var err error
+			if suppression.Canceled {
+				err = wb.Delete([]byte(key))
+			} else {
+				err = wb.Set([]byte(key), []byte(""))
+			}
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	if err := wb.Set([]byte(tokenKey), token); err != nil {
+		return err
+	}
+	err := wb.Flush()
+	return err
+}
+
+// start starts the repository
+func (b *Repository) start() error {
+	b.closed = make(chan struct{})
+	var seeder io.Reader
+	if _, err := os.Stat(b.path); os.IsNotExist(err) && b.seederSource != nil {
+		if seeder, err = b.seederSource(); err != nil {
+			return err
+		}
+	}
+
+	opts := badger.
+		DefaultOptions(b.path).
+		WithLogger(blogger{b.log}).
+		WithCompression(options.None).
+		WithIndexCacheSize(16 << 20).
+		WithNumGoroutines(b.maxGoroutines)
+	var err error
+	b.db, err = badger.Open(opts)
+	if err != nil {
+		return err
+	}
+
+	if seeder != nil {
+		restoreDone := lo.Async(func() error {
+			return b.Restore(seeder)
+		})
+		select {
+		case <-restoreDone:
+		case <-time.After(b.maxSeedWait):
+		}
+	}
+
+	go func() {
+		ticker := time.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-b.closed:
+				return
+			case <-ticker.C:
+			}
+		again:
+			err := b.db.RunValueLogGC(0.7)
+			if err == nil {
+				goto again
+			}
+		}
+	}()
+	return nil
+}
+
+// Stop stops the repository
+func (b *Repository) Stop() error {
+	var err error
+	b.closeOnce.Do(func() {
+		close(b.closed)
+		err = b.db.Close()
+	})
+	return err
+}
+
+// Backup writes a backup of the repository to the given writer
+func (b *Repository) Backup(w io.Writer) error {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return model.ErrRestoring
+	}
+	_, err := b.db.Backup(w, 0)
+	return err
+}
+
+// Restore restores the repository from the given reader
+func (b *Repository) Restore(r io.Reader) (err error) {
+	if b.isRestoring() {
+		return model.ErrRestoring
+	}
```

Comment:
This is probably not needed given that `Restore` is called only once in `start`, I don't see how we could have overlapping `Restore` calls.

Reply:
Restore is an exposed method and actually being used by tests & benchmarks